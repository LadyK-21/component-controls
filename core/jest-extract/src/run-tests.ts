import path from 'path';
import { runCLI } from 'jest';
import { Config } from '@jest/types';
import { findUpFile } from '@component-controls/core/node-utils';
import { log } from '@component-controls/logger';
import { TestResult, AggregatedResult } from '@jest/test-result';
import { FileCoverage, CoverageSummaryData } from 'istanbul-lib-coverage';
import fastq from 'fastq';

export const findJestConfig = (filePath: string): string =>
  path.dirname(
    findUpFile(path.dirname(filePath), [
      'package.json',
      'jest-config.js',
      'jest-config.ts',
    ]) || filePath,
  );

export interface JestResults {
  /**
   * test results
   */
  results: Pick<
    TestResult,
    'leaks' | 'memoryUsage' | 'perfStats' | 'testFilePath' | 'testResults'
  >[];

  /**
   * coverage summary data, by file
   */
  coverage: Record<string, CoverageSummaryData>;
}

interface TestWorkerInput {
  testFiles: string[];
  projectFolder: string;
  options: Partial<Config.Argv>;
}
const runTestsWorker: fastq.asyncWorker<
  unknown,
  TestWorkerInput,
  JestResults | undefined
> = async ({ testFiles, projectFolder, options }: TestWorkerInput) => {
  let runResults: {
    results: AggregatedResult;
    globalConfig: Config.GlobalConfig;
  };
  try {
    runResults = await runCLI(
      {
        rootDir: projectFolder,
        testRegex: testFiles,
        forceExit: true,
        testPathIgnorePatterns: ['/node_modules/', '/__snapshots__/'],
        silent: true,
        verbose: false,
        testTimeout: 100000,
        reporters: [],
        coverage: true,
        coverageReporters: ['none'],
        watchman: false,
        collectCoverageFrom: [
          '**/*.{js,jsx,tsx,ts}',
          '!**/jest.config.js',
          '!**/*.{test,spec}.{js,jsx,tsx,ts}',
        ],
        ...options,
      } as Config.Argv,
      [projectFolder],
    );
  } catch (err) {
    return undefined;
  }
  const { coverageMap, testResults } = runResults.results;
  const result: JestResults = {
    results: testResults.map(
      ({ leaks, memoryUsage, perfStats, testFilePath, testResults }) => ({
        leaks,
        memoryUsage,
        perfStats: { ...perfStats },
        testFilePath: path.relative(projectFolder, testFilePath),
        testResults: [...testResults],
      }),
    ),
    coverage: {},
  };
  if (coverageMap) {
    Object.keys(coverageMap.data).forEach(file => {
      const fc = coverageMap.data[file] as FileCoverage;
      const summary = fc.toSummary().toJSON();
      const totals = Object.values(summary).reduce(
        (total, value) => total + value.covered + value.skipped,
        0,
      );
      if (totals) {
        result.coverage[path.relative(projectFolder, file)] = summary;
      }
    });
  }
  return result;
};

let queue: fastq.queueAsPromised<
  TestWorkerInput,
  JestResults | undefined
> | null = null;

/**
 * single test file execution
 * @param testFilePath test file full path
 * @param options jest runCLI options
 * @returns jest test results and coverage
 */
export const runTests = async (
  testFilePath: string,
  options: Partial<Config.Argv> = {},
): Promise<JestResults> => {
  const testFiles = [path.basename(testFilePath)];
  const projectFolder = findJestConfig(testFilePath);

  return runProjectTests(testFiles, projectFolder, options);
};

/**
 * run tests within the same project (jest config file)
 * @param testFiles an array of test files to run
 * @param projectFolder fle path where the jest config/package.json file resides
 * @param options jest runCLI options
 * @returns jest test results and coverage
 */
export const runProjectTests = async (
  testFiles: string[],
  projectFolder: string,
  options: Partial<Config.Argv> = {},
): Promise<JestResults> => {
  if (!queue) {
    queue = fastq.promise(runTestsWorker, 1);
  }
  const result = await queue.push({ testFiles, projectFolder, options });
  const maxFilesReport = 2;
  const reportFiles = testFiles.slice(0, maxFilesReport);
  if (testFiles.length > maxFilesReport) {
    reportFiles.push(`...${testFiles.length - maxFilesReport} more`);
  }
  log(`tests ${projectFolder}`, reportFiles.join(', '));
  return (result as unknown) as JestResults;
};
