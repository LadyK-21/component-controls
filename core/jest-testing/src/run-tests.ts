import fs from 'fs';
import path from 'path';
import { runCLI } from 'jest';
import { Config } from '@jest/types';
import { AssertionResult, AggregatedResult } from '@jest/test-result';
import { FileCoverage, CoverageSummaryData } from 'istanbul-lib-coverage';

export const getUniqueFolder = (folder: string): string => {
  let fnId = 1;
  const fName = () => path.resolve(folder, `__jest-tmp-${fnId}`);
  while (fs.existsSync(fName())) {
    fnId += 1;
  }
  const folderName = fName();
  fs.mkdirSync(folderName);
  return folderName;
};
export interface JestResults {
  /**
   * test results
   */
  testResults: AssertionResult[];
  /**
   * coverage summary data, by file
   */
  coverage: Record<string, CoverageSummaryData>;
}
export const run = async (
  testFilePath: string,
  jestConfig: Partial<Config.Argv> = {},
): Promise<JestResults | undefined> => {
  const testFolder = path.dirname(testFilePath);
  const testFile = path.basename(testFilePath);
  const configFolder = getUniqueFolder(testFolder);
  const configFile = path.resolve(configFolder, 'jest.config.js');
  const { rootDir = '..', ...config } = jestConfig;
  fs.writeFileSync(
    configFile,
    `module.exports = ${JSON.stringify(
      {
        rootDir,
        preset: 'ts-jest',
        transform: {
          '^.+\\.(ts|tsx)?$': 'ts-jest',
          '^.+\\.(js|jsx)$': 'babel-jest',
        },
        collectCoverageFrom: [
          '**/*.{js,jsx,tsx,ts}',
          '!**/jest.config.js',
          '!**/*.{test,spec}.{js,jsx,tsx,ts}',
        ],
      },
      null,
      2,
    )}
  `,
    'utf8',
  );
  let runResults: {
    results: AggregatedResult;
    globalConfig: Config.GlobalConfig;
  };
  try {
    runResults = await runCLI(
      {
        testRegex: testFile,
        testPathIgnorePatterns: ['/node_modules/', '/__snapshots__/'],
        silent: true,
        verbose: false,
        reporters: [],
        coverage: true,
        coverageReporters: ['none'],
        watchman: false,
        ...config,
      } as Config.Argv,
      [configFolder],
    );
  } catch (err) {
    console.error(err);
    return undefined;
  } finally {
    fs.unlinkSync(configFile);
    fs.rmdirSync(configFolder, { recursive: true });
  }
  const cov = runResults.results.coverageMap;
  if (runResults.results.testResults.length !== 1) {
    throw `More than one test results returned ${runResults.results.testResults.length}`;
  }
  const result: JestResults = {
    testResults: runResults.results.testResults[0].testResults,
    coverage: {},
  };
  if (cov) {
    Object.keys(cov.data).forEach(file => {
      const fc = cov.data[file] as FileCoverage;
      const summary = fc.toSummary().toJSON();
      const totals = Object.values(summary).reduce(
        (total, value) => total + value.covered + value.skipped,
        0,
      );
      if (totals) {
        result.coverage[path.relative(testFolder, file)] = summary;
      }
    });
  }

  return result;
};
