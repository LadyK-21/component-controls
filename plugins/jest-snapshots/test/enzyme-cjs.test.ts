import path from 'path';
import { runTests } from './run-tests';

runTests({
  format: 'cjs',
  renderer: 'enzyme',
  configPath: path.resolve(__dirname, '.config'),
});
