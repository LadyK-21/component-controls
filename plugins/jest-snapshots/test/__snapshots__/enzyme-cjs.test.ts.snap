// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enzyme  cjs 1`] = `
"const path = require('path');
const {
  loadConfigurations,
  extractDocuments,
} = require('@component-controls/config');
const { renderExample } = require('@component-controls/test-renderers');
const { render: reactRender } = require('@component-controls/render/react');

const { mount, configure } = require('enzyme');
const toJson = require('enzyme-to-json').default;
const Adapter = require('@wojtekmaj/enzyme-adapter-react-17');

configure({ adapter: new Adapter() });

describe('component-controls generated', () => {
  beforeAll(async done => {
    jest.mock('rc-util/lib/Portal');
    global.MutationObserver = class {
      constructor() {}
      disconnect() {}
      observe() {}
      takeRecords() {
        return [];
      }
    };

    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
    done();
  });
  const configPath = path.resolve(__dirname, '.config');
  const config = loadConfigurations(configPath);
  if (!config.renderFn) {
    config.renderFn = reactRender;
  }
  const documents = extractDocuments({ config, configPath });
  if (documents) {
    documents.forEach(file => {
      const exports = require(file);
      const doc = exports.default;
      const examples = Object.keys(exports)
        .filter(key => key !== 'default')
        .map(key => exports[key]);

      if (examples.length) {
        describe(doc.title, () => {
          examples.forEach(example => {
            it(example.name, async () => {
              let rendered;
              rendered = renderExample({
                example,
                doc,
                config,
              });
              const component = mount(rendered);
              const serialize = () => toJson(component, { mode: 'deep' });

              expect(serialize()).toMatchSnapshot();
            });
          });
        });
      }
    });
  }
  
});
"
`;
